crates/bitwarden/src/crypto/master_key.rs

```diff
@@ -1,17 +1,59 @@
 use aes::cipher::typenum::U32;
+use base64::Engine;
 
-use super::hkdf_expand;
+use crate::util::BASE64_ENGINE;
+
+use super::{
+    hkdf_expand, EncString, PbkdfSha256Hmac, SymmetricCryptoKey, PBKDF_SHA256_HMAC_OUT_SIZE,
+};
 use {
     crate::{client::auth_settings::Kdf, error::Result},
     aes::cipher::generic_array::GenericArray,
     sha2::Digest,
 };
 
-pub(crate) fn hash_kdf(secret: &[u8], salt: &[u8], kdf: &Kdf) -> Result<[u8; 32]> {
-    use crate::crypto::PBKDF_SHA256_HMAC_OUT_SIZE;
+#[derive(Copy, Clone)]
+pub(crate) enum HashPurpose {
+    ServerAuthorization = 1,
+    // LocalAuthorization = 2,
+}
+
+/// A Master Key.
+pub(crate) struct MasterKey(SymmetricCryptoKey);
+
+impl MasterKey {
+    /// Derives a users master key from their password, email and KDF.
+    pub fn derive(password: &[u8], email: &[u8], kdf: &Kdf) -> Result<Self> {
+        derive_key(password, email, kdf).map(Self)
+    }
+
+    /// Derive the master key hash, used for server authorization.
+    pub(crate) fn derive_master_key_hash(
+        &self,
+        password: &[u8],
+        purpose: HashPurpose,
+    ) -> Result<String> {
+        let hash = pbkdf2::pbkdf2_array::<PbkdfSha256Hmac, PBKDF_SHA256_HMAC_OUT_SIZE>(
+            &self.0.key,
+            password,
+            purpose as u32,
+        )
+        .expect("hash is a valid fixed size");
+
+        Ok(BASE64_ENGINE.encode(hash))
+    }
 
-    use super::PbkdfSha256Hmac;
```

Comment:
The two feature flags on this file for "internal" can be removed, because it's already feature flagged in the mod file.